Use country-phone input component for terminal settings phone field

elazzabi opened the pull request:
Fixes #3986 

#### Changes proposed in this Pull Request

Use country-phone component instead of a normal input field.

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

No regression between the old component and the new one.

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

Before:

<img width="817" alt="Screen Shot 2022-06-16 at 16 44 18" src="https://user-images.githubusercontent.com/12542046/174111002-8c5ec1dc-e93b-4940-8c95-194ade3fcf5e.png">


After:

<img width="892" alt="Screen Shot 2022-06-16 at 16 17 12" src="https://user-images.githubusercontent.com/12542046/174104330-d1276dfc-6712-413f-9251-6d2366d1af4e.png">

-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**


<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

elazzabi commented:
⚠️ 
**Limitations**:
The component was designed for checkout and a lot of the decisions/code is tight to the checkout process: hard-coded class names, etc... I tried to keep it as it is, so I don't break the checkout process, but if we want to reuse that component in different places, it needs to be decoupled from the checkout process.

So far, the biggest limitation is the styling of the component and its interaction with elements on the page (hard-coded class names...)

kalessil commented:
@elazzabi : so we managed to make it work here without side effects but the component is not as reusable as might be?

elazzabi commented:
@kalessil Yes, it does work, but there are a few problems:
- we can't style it to make it look like the other inputs on the page (the width of the component is different from others on the page for example)
- it's still trying to interact with other inputs on the page via hardcoded classes names, but those do not exist (exists only in the checkout process)

kalessil commented on client/card-readers/settings/sections/test/index.test.js:
```diff
@@ -71,7 +71,7 @@ describe( 'Card Reader Contact Details section', () => {
 		expect( email ).toBeInTheDocument();
 
-		const phone = screen.getByLabelText( 'Support phone number' );
+		const phone = screen.getByLabelText( 'Mobile phone number' );
```
Is this change due to the component usage (I guess a hardcoded value there)?

elazzabi replied:
Yeah, the label `Mobile phone number` is hardcoded inside the component, so I can't pass one...

kalessil commented:
@elazzabi : IIRC, under the hood, it uses a third-party component. How much effort it would be to use the third-party one here directly?

elazzabi commented:
@kalessil good call. I believe most of the code added to the component is to make it work with the checkout process, so I don't think it will be hard to use the component directly. I'll double-check and update this PR accordingly.

kalessil commented:
Super, thanks @elazzabi !

elazzabi commented on client/components/platform-checkout/save-user/phone-number-input.js:
```diff
@@ -1,144 +1,111 @@
 
 			// Focus the phone number input when the component loads.
-			input.focus();
+			inputRef.current.focus();
```
I'm not sure why this is used. I believe we shouldn't automatically focus on the mobile input when the page is loaded, but maybe I'm missing something...

Should I just go and remove it?

elazzabi commented:
@kalessil @mordeth Finally made the component decoupled from that page. Changes introduced are:
1. Removed all hard-coded values, class names, or others. They are now either passed via props or taken care of by the parent component
1. Moved the error handling to parent component. This way we can display errors depending on the page-styling, etc...
1. Removed a bunch of internal state that can be replaced by the parent component. The PhoneInput component is now used for representation only + communicating if the phone is valid or not with the parent for styling/error-handling.

What's missing from this PR (I'll work on it next):
1. TypeScript support
2. Fixing tests (if any broken ones)
3. Running some manual tests for checkout phase

anu-rock commented:
Also invited Mayisha to review as she had originally added the tel input component in #3699.

@Mayisha - We are trying to adapt the tel input component on card reader settings page. We realized that the component was customized for checkout, so we took some steps to make it more generic. Do let us know your thoughts.

mordeth reviewed (commented):
Looks good so far! I have added few comments. Something I would change is the location of the component. As it is no longer Platfrom checkout specific, I think `settings` is much better place for it.

Mayisha commented on client/components/platform-checkout/save-user/checkout-page-save-user.js:
```diff
@@ -14,18 +14,50 @@ import AdditionalInformation from './additional-information';
+		}
+
+		const updateFormSubmitButton = () => {
+			if ( isPhoneValid ) {
```
`isPhoneValid` is `null` initially and it does not change until the user clicks the `Save my information for faster checkouts` checkbox or edits the phone number input field. For the `null` value the `Place Order` button is disabled though I have all the required fields filled up in the original checkout blocks.

![disabled](https://user-images.githubusercontent.com/33387139/175365824-ef24f911-633c-4de6-b96c-430afea6286a.png)

The button gets enabled as soon as I select the checkbox.

elazzabi replied:
Can you please let me know how to reproduce this? I can't reproduce it on my end:
<img width="874" alt="Screen Shot 2022-06-28 at 11 48 30" src="https://user-images.githubusercontent.com/12542046/176161160-7e58919e-18f0-4a95-bd60-5dec2060532d.png">

Mayisha replied:
To see the `Remember your details?` section, you need to have `WooPay` enabled on your merchant site. 

First, you need to set it up in WCPay server DB. Follow any one of the following steps-
- Run this command `wp wcpay enable_woopay <merchant-blog-id>` in your WCPay Server's container.

OR

- In WCPay Server's `wcpay_account_meta` table, add a meta key `platform_checkout_eligible` for your WCPay account, with a meta value of `1`. To find your WCPay account ID, look at the `wcpay_accounts` table.

Then on your merchant site UI enable **wp-admin > Payments > Settings > Express checkouts > WooPay**

![woopay](https://user-images.githubusercontent.com/33387139/176758067-0fe3467d-bb7c-44c9-8ccb-f1546e4b77d2.png)

Finally on the merchant's checkout page input an email id and select credit card as payment method. The checkbox will appear now. :tada:

Mayisha commented on client/components/platform-checkout/save-user/phone-number-input.js:
```diff
@@ -1,144 +1,111 @@
+	inputProps = {},
+} ) => {
 	const [ inputInstance, setInputInstance ] = useState( null );
 	const [ isValid, setIsValid ] = useState( true );
```
We now have 2 state variables indicating whether the phone number is valid or not, one in the parent component and another in this. How about we consolidate the variables and have just one?

Mayisha commented on client/components/platform-checkout/save-user/phone-number-input.js:
```diff
@@ -1,144 +1,111 @@
 
 		const handleCountryChange = () => {
-			handlePhoneNumberChange( iti.getNumber() );
+			onValueChange( iti.getNumber() );
```
Let's call the `onValidationChange( iti.isValidNumber() )` also here. Currently if I change the country from the error message does not appear even if it is an invalid phone number for the newly selected country.

Mayisha commented on client/components/platform-checkout/save-user/phone-number-input.js:
```diff
@@ -1,144 +1,111 @@
 			if ( iti ) {
 				iti.destroy();
-				input.removeEventListener(
+				inputRef.removeEventListener(
```
Small typo here, should be `inputRef.current`. This gives an error in the console currently.

Mayisha reviewed (changes requested):
I have tested the flow on the checkout page and found a couple of small issues that need to be fixed. :eyes: 
Apart from that, I agree with @mordeth about moving the component to a generic place as an independent component. :+1:

elazzabi commented on client/components/platform-checkout/save-user/checkout-page-save-user.js:
```diff
@@ -14,18 +14,50 @@ import AdditionalInformation from './additional-information';
 	} = useSelectedPaymentMethod();
 
+	// This runs once to populate the phone number state if already present in the DOM.
+	useEffect( () => {
+		setPhoneNumber(
+			document.getElementById( 'billing_phone' )?.value ?? ''
+		);
+	}, [] );
```
Do you mean the phone input field is not populated and we can get rid of this, or this gets fired before the population of the field? 🤔

Mayisha replied:
>  or this gets fired before the population of the field? 

Yes, this statement is right.

So now it's fired on the checkout page load only. Previously it used to fire on mount the phone input component. If I check/uncheck the `save user` checkbox the phone input component mounts/unmounts. 

Consider this flow on trunk-
- checkout page loads and the phone field (the one in the billing address section) is empty initially. The checkbox is also unchecked
- User adds a phone number
- User selects the checkbox
- The phone number input is populated. This happens because the phone number input component mounts on selecting the checkbox, the hook is triggered and takes the value of the `billing_phone`.

Now in this branch-
- checkout page loads and the phone field (the one in the billing address section) is empty initially. The checkbox is also unchecked. But on the checkbox component the hook is triggered and received empty value as `billing_phone`
- User adds a phone number
- User selects the checkbox
- The phone number input field is empty now because the hook does not get triggered on this child component mount and shows the empty string that was initially obtained from the `billing_phone` value on it's parent component.

So the behavior is deviated from how it is on the trunk. Does this make sense? :thinking:

elazzabi replied:
Thank you for the detailed steps, I can reproduce it now and I think it's fixed by https://github.com/Automattic/woocommerce-payments/pull/4355/commits/9b934c8939f9e893b96441dd9e72b9799e1cec08

elazzabi commented:
Thank you @Mayisha for the review! Sorry, it took me a while to respond as I was AFK.

I fixed the bugs you raised, and I asked a few questions for the remaining one. Thank you for your help!

Mayisha commented:
Thanks @elazzabi for addressing the issues :rocket: 

I have tested this again on the checkout page and all the mentioned issues have been addressed and working perfectly. :tada: 

I noticed one new issue, I can not clear the phone field up to the last digit now in order to input a completely new number. It would be great if this one is also handled :eyes: 

![can not clear](https://user-images.githubusercontent.com/33387139/177538774-3d7ebf3b-f115-4fb7-81a1-b8ab4b4287e5.gif)

elazzabi commented:
Thank you @Mayisha for the review and catching the issue! I think this one is fixed now

https://user-images.githubusercontent.com/12542046/177749408-53184b6b-cca0-4952-9478-b32e3b408cda.mov

Also:
- I switched the component to TypeScript
- I fixed tests and added a new one

elazzabi commented on client/settings/phone-input/index.tsx:
```diff
@@ -0,0 +1,127 @@
+const PhoneNumberInput = ( {
+	onValueChange,
+	value,
+	onValidationChange = ( validation ) => validation,
```
This technically could be an empty function, but TypeScript keeps complaining about an empty function

elazzabi commented on client/settings/phone-input/index.tsx:
```diff
@@ -0,0 +1,127 @@
+				ref={ inputRef }
+				value={ removeInternationalPrefix( value ) }
+				onChange={ handlePhoneNumberInputChange }
+				aria-label={
```
FYI: After using TypeScript I ditched the `label` attribute. It's not a correct HTML attr for the `input` element and it was raised by TypeScript. The correct one is `aria-label`

Mayisha commented on client/settings/phone-input/index.tsx:
```diff
@@ -0,0 +1,127 @@
+/**
+ * Internal dependencies
+ */
+// @ts-ignore: Cannot find module
```
Probably adding the path in `tsconfig.json` could help? :thinking:

mgascam replied:
It seems that this dependency is not using TS and the compiler is failing to find the type definition. There is [a way to declare it and make the compiler happy](https://www.typescriptlang.org/docs/handbook/modules.html#working-with-other-javascript-libraries).

You can add a `declarations.d.ts` in the folder of the component. In this file put the following:

`declare module 'iti/utils';`

Note that this is the shorthand notation for module declaration. Anything imported from that module will be considered to be of `any` type. You can also define the types that the module exports if you need. There are examples in the codebase you can check like [here](https://github.com/Automattic/woocommerce-payments/blob/a703cbd14f375bfa2bc93d8b1d1b3cdddd716ab8/client/components/deposits-information/declarations.d.ts)

elazzabi replied:
Thanks @mgascam ! It looks like it's a two step process, and the second part involves what you shared ^

So, first part was adding a `path` to tell TypeScript that `iti/utils` => `../node_modules/intl-tel-input/build/js/utils`.
The second part is declaring the module given no type is exported by the package.

Added here https://github.com/Automattic/woocommerce-payments/pull/4355/commits/cb980c3b3e7b8a1dc6baa8765d4255f1fc604860

mgascam commented on client/settings/phone-input/iti.d.ts:
```diff
@@ -0,0 +1 @@
+declare module 'iti/utils';
```
These files are usually named `declarations.d.ts` in the code base. Maybe you'll want to stick to that naming?

mgascam commented on client/settings/phone-input/index.tsx:
```diff
@@ -0,0 +1,125 @@
+			onValidationChange( iti.isValidNumber() );
+
+			// Focus the phone number input when the component loads.
+			currentRef.focus();
```
In the Card Readers page, when the form loads  the input phone component gets the focus. It feels a bit weird since the phone input is not the first in the form. Not sure how hard it can be to handle it in a different way, without giving the focus to the component?

elazzabi replied:
Oh yes, I agree. This was part of [another review](https://github.com/Automattic/woocommerce-payments/pull/4355#discussion_r903421420) but I missed it. Thank you for pointing it out!

mgascam commented on client/settings/phone-input/index.tsx:
```diff
@@ -0,0 +1,125 @@
+		let iti: intlTelInput.Plugin | null = null;
+		const currentRef = inputRef.current;
+
+		const handleCountryChange = () => {
```
I think I found a bug related to the country change. Steps to reproduce:

1. Go to Payments > Card Readers > Receipt details
2. Remove some characters from the input component so you see the invalid phone error
3. Click on the country selector and change the country
4. Observe the issue

**Actual behavior**
When the country changes the phone number is rendered incorrectly. It includes the phone prefix as part of the number. 
There is a warning thrown to the browser console:
`react_devtools_backend.js:4026 Warning: Maximum update depth exceeded. This can happen when a component calls setState inside useEffect, but useEffect either doesn't have a dependency array, or one of the dependencies changes on every render.
 `
In some ocassions I noticed that the page becomes unresponsive and eventually crashes. Unfortunately I could not find the exact combination of steps to reproduce, but I think it can be reproduced playing with the component (edit phone number and change country several times)

I recorded this [video](https://d.pr/v/RlpfR5) that shows the issue in action. Please let me know if I can give any further details

elazzabi replied:
Aha, weird bug. Thank you for pointing it out!

That doesn't happen on the Checkout page, so I believe it's something specific to the Card Readers page.

Looking...

mgascam replied:
Hey @elazzabi I think the problem could be related to the useEffect dependency array. Now we have:

```
useEffect( () => {
		// ...
	}, [ onValueChange, onValidationChange ] );
```
Since both dependencies are functions I suspect the reference changes on every render, causing the useEffect to run again. I used a different approach that seems to work, maybe you'd like to give it a try:

```
useEffect( () => {
		// ...
	}, [ inputRef ] );
```
With this approach, the effect will run if inputRef changes which I think is correct.

elazzabi replied:
Thank you for the pointers @mgascam! I took a look and, changing the dependencies of `useEffect` to `inputRef` fixes it. However, that fails the lint process given `useEffect` uses both `onValueChange` and `onValidationChange`, and thus, they need to appear there.

So, I left them both but looked deeper into what was causing the issue. It appears that `onValidationChange` is changing on every render, and thus makes `useEffect` re-run on every render.

I wrapped a few functions with `useCallback` to avoid the issue. Now it passes, but the new problem is the input loses focus on every key stroke 😅

elazzabi replied:
Ok, back to this.

The problem with this is that the input gets re-rendered on each keystroke. This is derived from a long chain of changes:

Changes to the input are tracked via the `accountBusinessSupportPhone` variable. Any change to this variable will trigger a re-calculation of the updateSaveButtonAvailability [function](https://github.com/Automattic/woocommerce-payments/pull/4355/commits/a8b78b0adc91c15a6d5d85cb65f30799e1a83826#diff-606705683552d06907190d85b12172ed4482010bf9f80f5c39ebc94636458055R67). Each change to this function will trigger a re-calculation of the `onValueChange` [prop](https://github.com/Automattic/woocommerce-payments/pull/4355/commits/a8b78b0adc91c15a6d5d85cb65f30799e1a83826#diff-606705683552d06907190d85b12172ed4482010bf9f80f5c39ebc94636458055R107). And finally, each change to the prop will re-trigger the `useEffect` in the input component. Inside the useEffect, we attach the library to the input. So every time we execute it, the library is re-attached and the focus is lost.

The real problem is I made the `onValueChange` make so many things, which is not ideal, my bad. Especially, [the part where it checks if the Save button should be enabled](https://github.com/Automattic/woocommerce-payments/pull/4355/commits/75602c6dc37f9063349da4d104e3ab99c64081b0#diff-606705683552d06907190d85b12172ed4482010bf9f80f5c39ebc94636458055L105). Now, it is fixed to only change the value. Checking for the button enable/disable state is [happening outside](https://github.com/Automattic/woocommerce-payments/pull/4355/commits/75602c6dc37f9063349da4d104e3ab99c64081b0#diff-606705683552d06907190d85b12172ed4482010bf9f80f5c39ebc94636458055R59). This has the benefit of being independent and will be used by [other inputs too](https://github.com/Automattic/woocommerce-payments/pull/4355/commits/75602c6dc37f9063349da4d104e3ab99c64081b0#r919136794).

mgascam replied:
Nice! I am not able to reproduce the bug anymore, well done

elazzabi commented on client/card-readers/settings/sections/contacts-details.js:
```diff
@@ -69,21 +85,28 @@ const ContactDetailsSection = ( { setSaveDisabled } ) => {
 				value={ accountBusinessSupportEmail }
 				onChange={ setAccountBusinessSupportEmail }
 				type="email"
-				onBlur={ updateSaveButtonAvailability }
```
This is moved from here to the `useEffect` ^

mgascam commented on client/card-readers/settings/sections/contacts-details.js:
```diff
@@ -69,21 +85,28 @@ const ContactDetailsSection = ( { setSaveDisabled } ) => {
-				onBlur={ updateSaveButtonAvailability }
-			/>
+			<div className="card-readers-business-phone-input">
+				<span>
```
I guess this could be `label`. I noticed there is less margin between text and the input:

<img width="411" alt="Screen Shot 2022-07-13 at 10 44 27" src="https://user-images.githubusercontent.com/1553182/178691881-62f0e872-bdfc-4d68-8e7b-77000265d221.png">

elazzabi replied:
Oh, missed this!

I wanted to hard-code a margin, but then made a quick search and found the `BaseControl` exposed from `@wordpress/components`. I ended up wrapping the component with it https://github.com/Automattic/woocommerce-payments/pull/4355/commits/b331184c9a2f5367faf0668906bffcf4b48362ad

mgascam reviewed (approved):
LGTM! Tested the new component on Card Readers settings page and WooPay checkout ✅

mordeth reviewed (approved):
LGTM! ✅
